Source/Core/VideoCommon/TextureConversionShader.cpp

```diff
@@ -720,4 +724,526 @@ const char* GenerateEncodingShader(u32 format, APIType ApiType)
   return text;
 }
 
+// NOTE: In these uniforms, a row refers to a row of blocks, not texels.
+static const char decoding_shader_header[] = R"(
+#ifdef VULKAN
+
+layout(std140, push_constant) uniform PushConstants {
+  uvec2 dst_size;
+  uvec2 src_size;
+  uint src_offset;
+  uint src_row_stride;
+  uint palette_offset;
+} push_constants;
+#define u_dst_size (push_constants.dst_size)
+#define u_src_size (push_constants.src_size)
+#define u_src_offset (push_constants.src_offset)
+#define u_src_row_stride (push_constants.src_row_stride)
+#define u_palette_offset (push_constants.palette_offset)
+
+TEXEL_BUFFER_BINDING(0) uniform usamplerBuffer s_input_buffer;
+TEXEL_BUFFER_BINDING(1) uniform usamplerBuffer s_palette_buffer;
+
+IMAGE_BINDING(rgba8, 0) uniform writeonly image2DArray output_image;
+
+#else
+
+uniform uvec2 u_dst_size;
+uniform uvec2 u_src_size;
+uniform uint u_src_offset;
+uniform uint u_src_row_stride;
+uniform uint u_palette_offset;
+
+SAMPLER_BINDING(9) uniform usamplerBuffer s_input_buffer;
+SAMPLER_BINDING(10) uniform usamplerBuffer s_palette_buffer;
+
+layout(rgba8, binding = 0) uniform writeonly image2DArray output_image;
+
+#endif
+
+uint Convert3To8(uint v)
+{
+  // Swizzle bits: 00000123 -> 12312312
+  return (v << 5) | (v << 2) | (v >> 1);
+}
+uint Convert4To8(uint v)
+{
+  // Swizzle bits: 00001234 -> 12341234
+  return (v << 4) | v;
+}
+uint Convert5To8(uint v)
+{
+  // Swizzle bits: 00012345 -> 12345123
+  return (v << 3) | (v >> 2);
+}
+uint Convert6To8(uint v)
+{
+  // Swizzle bits: 00123456 -> 12345612
+  return (v << 2) | (v >> 4);
+}
+
+uint GetTiledTexelOffset(uvec2 block_size, uvec2 coords)
+{
+  uvec2 block = coords / block_size;
+  uvec2 offset = coords % block_size;
+  uint buffer_pos = u_src_offset;
+  buffer_pos += block.y * u_src_row_stride;
+  buffer_pos += block.x * (block_size.x * block_size.y);
+  buffer_pos += offset.y * block_size.x;
+  buffer_pos += offset.x;
+  return buffer_pos;
+}
+
+uvec4 GetPaletteColor(uint index)
+{
+  // Fetch and swap BE to LE
+  uint val = texelFetch(s_palette_buffer, int(u_palette_offset + index)).x;
+  val = ((val >> 8) | (val << 8)) & 0xFFFFu;
+
+  uvec4 color;
+#if defined(PALETTE_FORMAT_IA8)
+  uint a = bitfieldExtract(val, 0, 8);
+  uint i = bitfieldExtract(val, 8, 8);
+  color = uvec4(i, i, i, a);
+#elif defined(PALETTE_FORMAT_RGB565)
+  color.x = Convert5To8(bitfieldExtract(val, 11, 5));
+  color.y = Convert6To8(bitfieldExtract(val, 5, 6));
+  color.z = Convert5To8(val & 0x1Fu);
```

Comment:
What do you think about using bitfieldExtract for all values? The compiler should parse it just fine.